File: dotNetTips.Utility/Extensions/DateTimeExtensions.vb
Imports System.Runtime.CompilerServices

Public Module DateTimeExtensions
    ''' <summary>
    ''' Returns the next occurrence of the day of the week specified.
    ''' </summary>
    ''' <param name="dt">The dt.</param>
    ''' <param name="dayOfWeek">The day of week.</param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    <Extension()>
    Public Function GetNext(ByVal dt As DateTime, ByVal dayOfWeek As DayOfWeek) As DateTime
        Dim daysToAdd As Integer = 0
        If dt.DayOfWeek < dayOfWeek Then
            daysToAdd = dayOfWeek - dt.DayOfWeek
        Else
            daysToAdd = (7 - dt.DayOfWeek) + dayOfWeek
        End If
        Return dt.AddDays(daysToAdd)
    End Function

    ''' <summary>
    ''' Gets the prior occurrence of the day of the week specified.
    ''' </summary>
    ''' <param name="dt">The dt.</param>
    ''' <param name="dayOfWeek">The day of week.</param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    <Extension()>
    Public Function GetLast(ByVal dt As DateTime, ByVal dayOfWeek As DayOfWeek) As DateTime
        Dim daysToSubtract As Integer = 0
        If dt.DayOfWeek > dayOfWeek Then
            daysToSubtract = dt.DayOfWeek - dayOfWeek
        Else
            daysToSubtract = (7 - dayOfWeek) + dt.DayOfWeek
        End If
        Return dt.AddDays(daysToSubtract * -1)
    End Function

    ''' <summary>
    ''' To the friendly date string.
    ''' </summary>
    ''' <param name="Date">The date.</param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    <System.Runtime.CompilerServices.Extension>
    Public Function ToFriendlyDateString([Date] As DateTime) As String
        Dim FormattedDate As String = ""
        If [Date].[Date] = DateTime.Today Then
            FormattedDate = "Today"
        ElseIf [Date].[Date] = DateTime.Today.AddDays(-1) Then
            FormattedDate = "Yesterday"
        ElseIf [Date].[Date] > DateTime.Today.AddDays(-6) Then
            ' *** Show the Day of the week
            FormattedDate = [Date].ToString("dddd").ToString()
        Else
            FormattedDate = [Date].ToString("MMMM dd, yyyy")
        End If

        'append the time portion to the output
        FormattedDate += " @ " + [Date].ToString("t").ToLower()
        Return FormattedDate
    End Function
End Module

File: dotNetTips.Utility.Standard.bak2/Extensions/DateTimeExtensions.cs
// ***********************************************************************
// Assembly         : dotNetTips.Utility.Standard
// Author           : David McCarter
// Created          : 01-22-2017
//
// Last Modified By : David McCarter
// Last Modified On : 01-22-2017
// ***********************************************************************
// <copyright file="DateTimeExtensions.cs" company="dotNetTips.Utility.Standard">
//     Copyright (c) dotNetTips.com - McCarter Consulting. All rights reserved.
// </copyright>
// <summary></summary>
// ***********************************************************************
using System;
using System.Globalization;

/// <summary>
/// The Extensions namespace.
/// </summary>
namespace dotNetTips.Utility.Standard.Extensions
{
    /// <summary>
    /// Extensions for DateTime
    /// </summary>
    public static class DateTimeExtensions
    {
        #region Public Methods

        /// <summary>
        /// Gets the last.
        /// </summary>
        /// <param name="input">The date/ time.</param>
        /// <param name="dayOfWeek">The day of week.</param>
        /// <returns>DateTime.</returns>
        /// <exception cref="ArgumentNullException">input - Input is invalid.</exception>
        public static DateTime GetLast(this DateTime input, DayOfWeek dayOfWeek)
        {
            if (Enum.IsDefined(typeof(DayOfWeek), input) == false)
            {
                throw new ArgumentNullException(nameof(input), "Input is invalid.");
            }

            var daysToSubtract = input.DayOfWeek > dayOfWeek ? input.DayOfWeek - dayOfWeek : (7 - (int)dayOfWeek) + (int)input.DayOfWeek;
            return input.AddDays(daysToSubtract * -1);
        }

        /// <summary>
        /// Gets the next.
        /// </summary>
        /// <param name="input">The date/ time.</param>
        /// <param name="dayOfWeek">The day of week.</param>
        /// <returns>DateTime.</returns>
        /// <exception cref="ArgumentNullException">input - Input is invalid.</exception>
        public static DateTime GetNext(this DateTime input, DayOfWeek dayOfWeek)
        {
            if (Enum.IsDefined(typeof(DayOfWeek), input) == false)
            {
                throw new ArgumentNullException(nameof(input), "Input is invalid.");
            }

            var daysToAdd = 0;
            daysToAdd = input.DayOfWeek < dayOfWeek ? dayOfWeek - input.DayOfWeek : (7 - (int)input.DayOfWeek) + (int)dayOfWeek;
            return input.AddDays(daysToAdd);
        }

        /// <summary>
        /// Intersects the specified start date.
        /// </summary>
        /// <param name="startDate">The start date.</param>
        /// <param name="endDate">The end date.</param>
        /// <param name="intersectingStartDate">The intersecting start date.</param>
        /// <param name="intersectingEndDate">The intersecting end date.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise.</returns>
        /// <remarks>Code by: Walter Quesada</remarks>
        public static bool Intersects(this DateTime startDate, DateTime endDate, DateTime intersectingStartDate, DateTime intersectingEndDate)
        {
            return intersectingEndDate >= startDate && intersectingStartDate <= endDate;
        }

        /// <summary>
        /// To the friendly date string.
        /// </summary>
        /// <param name="input">The date.</param>
        /// <returns>System.String.</returns>
        public static string ToFriendlyDateString(this DateTime input)
        {
            var formattedDate = string.Empty;

            if (input.Date == DateTime.Today)
            {
                formattedDate = nameof(DateTime.Today);
            }
            else
            {
                formattedDate = input.Date == DateTime.Today.AddDays(-1) ? Properties.Resources.Yesterday : input.Date > DateTime.Today.AddDays(-6) ? input.ToString("dddd", CultureInfo.CurrentCulture).ToString() : input.ToString(CultureInfo.CurrentCulture.DateTimeFormat.LongDatePattern, CultureInfo.CurrentCulture);
            }

            // append the time portion to the output
            formattedDate += " @ " + input.ToString(CultureInfo.CurrentCulture.DateTimeFormat.LongTimePattern, CultureInfo.CurrentCulture).ToLower();

            return formattedDate;
        }

        #endregion Public Methods
    }
}